src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -97,11 +97,20 @@ private BoundExpression RewriteDeconstruction(BoundTupleExpression left, Convers
                 builder.Add(resultPart);
             }
 
-            var tupleType = TupleTypeSymbol.Create(locationOpt: null, elementTypes: builder.SelectAsArray(e => e.Type),
-                elementLocations: default(ImmutableArray<Location>), elementNames: default(ImmutableArray<string>),
-                compilation: _compilation, shouldCheckConstraints: false, errorPositions: default(ImmutableArray<bool>));
+            if (isUsed)
+            {
+                var tupleType = TupleTypeSymbol.Create(locationOpt: null, elementTypes: builder.SelectAsArray(e => e.Type),
+                    elementLocations: default, elementNames: default,
+                    compilation: _compilation, shouldCheckConstraints: false, errorPositions: default);
 
-            return new BoundTupleLiteral(right.Syntax, default(ImmutableArray<string>), default(ImmutableArray<bool>), builder.ToImmutableAndFree(), tupleType);
+                return new BoundTupleLiteral(right.Syntax, argumentNamesOpt: default, inferredNamesOpt: default,
+                    arguments: builder.ToImmutableAndFree(), type: tupleType);
+            }
+            else
+            {
+                // When a deconstruction is not used, we use the last value of the LHS as the return value for the lowered form
+                return builder.ToImmutableAndFree().Last();
```

Comment:
>builder.ToImmutableAndFree().Last(); [](start = 23, length = 36)

Is it safe to drop content of the builder? It feels wasteful to allocate an array just to take the last item. Also, if we can drop all content of this builder, why are we using its item as the result. I think we should use the last assignment from effects.assignments, removing it from the list at the same time. Can we do that? It looks like we should be careful with nested case.